Add filter for public IPs on GossipSub peer exchange

cortze opened the pull request:
# Description
Following up on the issue raised in #617, this PR adds a new argument to the `WithPeerExchange` GossipSub option to restrict peer exchange to only include peer signed records with public IPs.

I can see a scenario where, despite a peer having a public IP in the PeerStore, we might not have a signed record for that IP. However, I think sharing just the PeerID (without private records) is still sufficient or at least better than sharing unsigned or private addresses.

I’m happy to discuss any of these points.

Kubuxu commented on gossipsub.go:
```diff
@@ -381,14 +382,15 @@ func WithFloodPublish(floodPublish bool) Option {
 // This should generally be enabled in bootstrappers and well connected/trusted nodes
 // used for bootstrapping.
-func WithPeerExchange(doPX bool) Option {
+func WithPeerExchange(doPX bool, pubMaddrsOnly bool) Option {
```
I would make this a separate option. This way it won't break the interface for everyone already using `WithPerrExchange`

cortze replied:
That makes sense. Updated on [69d3262](https://github.com/libp2p/go-libp2p-pubsub/pull/640/commits/69d326205c712ba810396ce02fb4e871d536c6c1)

Kubuxu commented on gossipsub.go:
```diff
@@ -389,6 +390,24 @@ func WithPeerExchange(doPX bool) Option {
+// WithPeerExchange is a gossipsub router option that enables Peer eXchange on PRUNE.
+// This should generally be enabled in bootstrappers and well connected/trusted nodes
+// used for bootstrapping but using only Signer PeerRecords that include public addreses.
+func WithPublicAddrsOnPeerExchange(doPX bool, pubMaddrsOnly bool) Option {
```
I would make this manipulate only `pubPX`, and `WithPeerExchange` shouldn't manipulate `pubPX`. This way there is no order dependency between these options.

Kubuxu commented:
It would be great if someone with more current libp2p experience also looked at it.

MarcoPolo reviewed (changes requested):
Hey @cortze thanks for the changes. Overall looks good, just a couple notes. you can find my comments in the last commit.

I've squashed all your commits into a single commit and made some changes that you can see here: https://github.com/libp2p/go-libp2p-pubsub/compare/7cbe1cbee004afa9ef43f7a3369ca1cd9d751ef7..83e3c602debcb09f5f73b5a2a7ae2fa8be0ee45f

I don't think we should send empty peer records. Happy to discuss further if you feel strongly.

Please squash your future changes and drop my review commit, thanks!

cortze commented:
Hey @MarcoPolo, thanks a lot for the review and the suggestions. 

I've addressed all the points except for the one regarding the empty peer records. My original thought was to follow what is described in the comment in [line 2015](https://github.com/libp2p/go-libp2p-pubsub/blob/d53e38b98dc291c355671e19400abb478ecb6e52/gossipsub.go#L2015) - where we would also ignore the signed records that are not good up to our public IP check:
> // see if we have a signed peer record to send back; if we don't, just send
> // the peer ID and let the pruned peer find them in the DHT -- we can't trust
> // unsigned address records through px anyway.

However, I don't have any strong opinion either. My only comment on this would be that if we were to send the SignedRecords without a public address anyway, we might be at a similar point as when the PR was started. 
The case that comes to my mind, and that I would like to avoid, is: 
- A node behind a NAT that connects our node
- We know it has a public IP and thus is in our Peerstore, but we didn't receive that address in the SignedPeerRecords.
- When checking the address we know about the peer, we see that the peer has a public IP
- However, that same IP is not within the SignedPeerRecord
I think it would still be okay and useful to share only the ID of the peer. The rest of the peers can use the ID to resolve the IP somehow without trying to dial the private addresses.

Either way, this is a good moment to choose the behaviour we think is best for the existing cases.

Also, and somehow out of the discussion, after talking with @dennis-tra, he raised a good point on making the GossipSub option open to any filtering function, not only the check for public addresses. I would make things easier to configure, but I would also like to hear what you think about this.

MarcoPolo commented:
> I’ve addressed all the points except for the one regarding the empty peer records. My original thought was to follow what is described in the comment in [line 2015](https://github.com/libp2p/go-libp2p-pubsub/blob/d53e38b98dc291c355671e19400abb478ecb6e52/gossipsub.go#L2015) - where we would also ignore the signed records that are not good up to our public IP check:

Ah, that’s a good point. Although the difference might be that the peer *didn’t* give us a signed peer record versus having given us one and it not containing a public address. The latter reveals a bit more about the peer, namely that the peer was capable of creating a signed peer record and chose not to include the public address.

> > // see if we have a signed peer record to send back; if we don’t, just send // the peer ID and let the pruned peer find them in the DHT – we can’t trust // unsigned address records through px anyway.
>
> However, I don’t have any strong opinion either. My only comment on this would be that if we were to send the SignedRecords without a public address anyway, we might be at a similar point as when the PR was started.

Agreed. We should never send signed peer records that contain no public addresses.

> The case that comes to my mind, and that I would like to avoid, is:
>
> - A node behind a NAT that connects our node
> - We know it has a public IP and thus is in our Peerstore, but we didn’t receive that address in the SignedPeerRecords.
> - When checking the address we know about the peer, we see that the peer has a public IP
> - However, that same IP is not within the SignedPeerRecord

In this case there may be no point in sharing the peer ID at all. The public IP address that you see is a NAT’ed IP address that won’t work for inbound connections. The only case it would be useful is if the node had registered a relay address that it shared on the DHT but not in its signed peer record.

> I think it would still be okay and useful to share only the ID of the peer. The rest of the peers can use the ID to resolve the IP somehow without trying to dial the private addresses.
>
> Either way, this is a good moment to choose the behaviour we think is best for the existing cases.

Maybe the safest/most-consistent-with-previous-behavior option is to include the peer id without the signed peer record as you initially did.

> Also, and somehow out of the discussion, after talking with @dennis-tra, he raised a good point on making the GossipSub option open to any filtering function, not only the check for public addresses. I would make things easier to configure, but I would also like to hear what you think about this.

I considered this as well. We recently merged a similar change in go-libp2p <https://github.com/libp2p/go-libp2p/commit/95cad5091fa1ee8af8b3cdf45c21630ca6b25dab>.

I think it’s a good idea. And it gives users an alternative from whatever we decide here. Let’s make it configurable with an easy option to enable what you propose in this PR (including just sending the PeerID in case of a signed peer record without public addresses).

Instead of `.pubPX` we’d have a `reducePXRecords` that takes a func of type: `func(pxRecords []*pb.PeerInfo, p peer.ID, addrs []multiaddr.Multiaddr, record *record.Envelope) []*pb.PeerInfo`

The pxPeers loop becomes:

``` go
for _, p := range pxPeers {
  var record *record.Envelope
  if cabOK {
    record = cab.GetPeerRecord(p)
  }
  px = reducePXRecords(px, p, gs.p.host.Peerstore().Addrs(p), record)
}
```

We’d have the following options:

``` go
func OnlyPublicAddrsOnPeerExchange() Option

type PXRecordReducer func(pxRecords []*pb.PeerInfo, p peer.ID, addrs []multiaddr.Multiaddr, record *record.Envelope) []*pb.PeerInfo

// WithCustomPXRecordReducer enables custom filtering of PX Records. See
// the implementation of OnlyPublicAddrsOnPeerExchange for an example.
func WithCustomPXRecordReducer(f PXRecordReducer) Option
```

The `OnlyPublicAddrsOnPeerExchange` option would use something like the following reducer:

``` go
func publicAddrsPXRecordReducer(pxRecords []*pb.PeerInfo, p peer.ID, addrs []multiaddr.Multiaddr, record *record.Envelope) []*pb.PeerInfo {
  if !slices.ContainsFunc(addrs, manet.IsPublicAddr) {
    return pxRecords
  }

  var recordBytes []byte
  if record != nil {
    err := checkPubAddresOnSignedPeerRecord(p, record);
    if err != nil {
      log.Warnf("no public address on signed peer record for %s: %s", p, err)
      // This peer has a public address, it just doesn't
      // have it in its signed peer record.
      // We'll share an empty record to at least inform the peer id.

    } else {
      var err error
      recordBytes, err = record.Marshal()
      if err != nil {
        log.Warnf("error marshaling signed peer record for %s: %s", p, err)
        return pxRecords
      }
    }
  }
  return append(pxRecords, &pb.PeerInfo{PeerID: []byte(p), SignedPeerRecord: recordBytes})
}
```

This probably also cleans up some complexity in the current makePrune method which is appreciated.

What do you think @cortze? If you agree, do you think you could update the PR?

cortze commented:
Sure! Let me make a second iteration over the PR over the day, and I'll squash the changes again.

MarcoPolo commented:
Looks good @cortze! I pushed a small change. Can you add a test to ensure we exchange the correct peer exchange records?

After that I think we're good here and can merge

cortze commented:
I've spotted a small issue on the `publicAddrsPXRecordReducer()` that was breaking the forwarding of the `peer.ID` as we agreed:
 
> Maybe the safest/most-consistent-with-previous-behavior option is to include the peer id without the signed peer record as you initially did.

Plus an extra check on `checkPubAddrsOnSignedPeerRecord()` that prevents panics with `nil` `*record.Envelop`.

Posting here the `git diff` of the `gossipsub.go` to make it easier to track:
```git
--- a/gossipsub.go
+++ b/gossipsub.go
@@ -444,7 +444,7 @@ func OnlyPublicAddrsOnPeerExchange() Option {
 func publicAddrsPXRecordReducer(pxRecords []*pb.PeerInfo, p peer.ID, addrs []multiaddr.Multiaddr, record *record.Envelope) []*pb.PeerInfo {
        if !slices.ContainsFunc(addrs, manet.IsPublicAddr) {
-               return pxRecords
+               return defaultPXRecordReducer(pxRecords, p, addrs, nil)
        }
        if !checkPubAddrsOnSignedPeerRecord(p, record) {
                log.Warnf("no public address on signed peer record for %s", p)

@@ -454,6 +454,9 @@ func publicAddrsPXRecordReducer(pxRecords []*pb.PeerInfo, p peer.ID, addrs []mul
 }

 func checkPubAddrsOnSignedPeerRecord(p peer.ID, env *record.Envelope) bool {
+       if env == nil {
+               return false
+       }
        rec, err := env.Record()
        if err != nil {
                log.Warnf("error getting peer record for %s from signed envelop: %v", p, err)
```

Thanks again @MarcoPolo for the feedback ✌🏽

cortze committed:
feat: add px peer record reducer + pub addrs filter

MarcoPolo commented:
> I've spotted a small issue on the `publicAddrsPXRecordReducer()` that was breaking the forwarding of the `peer.ID` as we agreed:
> 
> > Maybe the safest/most-consistent-with-previous-behavior option is to include the peer id without the signed peer record as you initially did.
> 
> Plus an extra check on `checkPubAddrsOnSignedPeerRecord()` that prevents panics with `nil` `*record.Envelop`

Nice, thanks!

> Posting here the `git diff` of the `gossipsub.go` to make it easier to track:

git range-diff is great for this: https://git-scm.com/docs/git-range-diff. e.g. `git range-diff master 68fb2e5dafa4812a924a95869df35ae19a51c88c 1e59ebd7b399fe2bbd765dbf4e18509599c8185d`

MarcoPolo commented:
> Looks good @cortze! I pushed a small change. Can you add a test to ensure we exchange the correct peer exchange records?
> 
> After that I think we're good here and can merge

Sorry I wasn't totally clear. I meant a end to end test with gossipsub and libp2p nodes to make sure the whole setup works as expected. If you don't get around to this, I'll add a test soon.

Thanks!

MarcoPolo commented:
Actually, we already have a test for PX in `TestGossipsubStarTopology`. So we're good here! I'll merge this after tests pass.
